packages/babel-traverse/src/path/inference/inferer-reference.js

```diff
@@ -91,9 +91,15 @@ function getTypeAnnotationBindingConstantViolations(binding, path, name) {
     }
   }
 
-  if (types.length) {
-    return t.createUnionTypeAnnotation(types);
+  if (!types.length) {
+    return;
+  }
+
+  if (types.every(t.isTSTypeAnnotation)) {
```

Comment:
It suffices to check via `types.some` since the TS type annotation will not be mixed with flow type annotation.

Reply:
Yeah, this is right
Fixed this ✅